docs/specification.rst

```diff
@@ -0,0 +1,217 @@
+++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+.. code-block:: text
+
+    Identifier ::= [a-zA-Z] [0-9a-zA-Z]*
+    Feature ::= Identifier ( '.' Identifier )*
+    S ::= #x9 | #x20
+    
+    DecimalValue ::= [0-9_]*
+    HexidecimalValue ::= [0-9a-fA-F_]+
+    OctalValue ::= [0-7_]+
+    BinaryValue ::= [01_]+
+    
+    VerilogValue ::= (( DecimalValue? S* "'" ( 'h' S* HexidecimalValue | 'b' S* BinaryValue | 'd' S*  DecimalValue | 'o' S* OctalValue ) | DecimalValue )
+    
+    FeatureAddress ::= '[' DecimalValue (':' DecimalValue)? ']'
+    
+    Any ::= [^#xA#]
+    Comment ::= '#' Any*
+    
+    AnnotationName ::= [.a-zA-Z] [_0-9a-zA-Z]*
+    NonEscapeCharacters ::= [^\"]
+    EscapeSequences ::= '\\' | '\"'
+    Annotation ::= AnnotationName S* '=' S* '"' (NonEscapeCharacters | EscapeSequences)* '"'
+    Annotations ::= '{' S* Annotation ( ',' S* Annotation )* S* '}'
+    
+    SetFasmFeature ::= Feature FeatureAddress? S* ('=' S* VerilogValue)?
+    FasmLine ::= S* SetFasmFeature? S* Annotations? S* Comment?
+
+Canonicalization
+++++++++++++++++
+
+If two FASM files have been canonicalized, then they enable an identical set of features. The canonical FASM file is also equivalent to the FASM file that would be generated by taking the output bitstream and converting it back into a FASM file.
+
+The canonicalization process is as follows:
+
+#. Flatten any ``FeatureAddress`` with width greater than 1.
+
+   * For ``SetFasmFeature`` lines with a ``FeatureAddress`` width greater than 1 bit, 1 ``SetFasmFeature`` for the width the original ``FeatureAddress``.
+   * When flattening, if the flattened address is 0, do not emit the address.
+#. Remove all comments and annotations.
+#. If the ``FeatureValue`` is 0, remove the FASM line.
+#. If the ``FeatureValue`` is 1, only output the ``Feature`` and the ``FeatureAddress`` if the ``Feature`` has addresses other than 0.
+#. Remove any lines that do not modify the default bitstream.
+
+   * Example are psuedo-pips in Xilinx parts.
+#. Sort the lines in the FASM file.
+
+Example Canonicalization
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+``ALUT.INIT[0] = 1``
+
+becomes
+
+``ALUT.INIT``
+
+----
+
+``ALUT.SMALL = 1``
+
+becomes
+
+``ALUT.SMALL``
+
+----
+
+``ALUT.INIT[3:0] = 4'b1101``
+
+becomes
+
+``ALUT.INIT``
+
+``ALUT.INIT[2]``
+
+``ALUT.INIT[3]``
+
+Meaning of a FASM line
+----------------------
+
+.. csv-table:: Simplified ``SetFasmFeature``
+    :delim: |
+    :header-rows: 1
+    
+    YYYY.XXXXX   | [A:B]              | = C
+    ``Feature``  | ``FeatureAddress`` | ``FeatureValue``
+    **Required** | *Optional*         | *Optional*
+
+Each line of a FASM file that enables a feature is defined by a ``SetFasmFeature``. Table 1 provides a simplified ``SetFasmFeature`` parse has three parts, the feature to be set (``Feature``), the address within the feature to be set (``FeatureAddress``) and the value of the feature (``FeatureValue``). Both the ``FeatureAddress`` and ``FeatureValue`` are optional.
```

Comment:
Table 1 provides a simplified version of ``SetFasmFeature`` parse.  A ``SetFasmFeature`` parse has three parts ...

Reply:
Done.